flashinfer/comm/allreduce.py

```diff
@@ -0,0 +1,746 @@
+
+    Hard requirements:
+    - SM90+ compute capability (checked by decorator)
+    - Multi-node topology
+    - Module availability
+    """
+    # MNNVL is designed for multi-node
+    if topology == "single_node":
+        return False
+
+    return True
+
+
+# ============================================================================
+# HEURISTIC - Performance-based selection for decorator
+# ============================================================================
+
+
+def _workspace_creation_heuristic(
+    suitable_backends: list[str],
+    backend: str,
+    world_size: int,
+    rank: int,
+    max_token_num: int,
+    hidden_dim: int,
+    dtype: torch.dtype,
+    topology: str,
+    **kwargs,
+) -> list[str]:
+    """
+    Select best backend for workspace creation based on performance.
+
+    Called by decorator after checking which backends pass requirements.
+    Uses benchmarking data to pick fastest option.
+
+    Args:
+        suitable_backends: List of backends that passed hard requirement checks
+        backend: Requested backend ("auto", "trtllm", or "mnnvl")
+        world_size: Number of ranks
+        rank: Current rank
+        max_token_num: Maximum number of tokens
+        hidden_dim: Hidden dimension size
+        dtype: Data type
+        topology: Network topology ("single_node" or "multi_node")
+        **kwargs: Additional arguments
+
+    Returns:
+        List containing the selected backend (single element)
+    """
+    if not suitable_backends:
+        return []
+
+    if len(suitable_backends) == 1:
+        return suitable_backends
+
+    # Decision tree based on benchmark data
+    # TODO: Replace with actual benchmarking results
+
+    # Multi-node: MNNVL is designed for this
+    if topology == "multi_node":
+        if "mnnvl" in suitable_backends:
+            return ["mnnvl"]
+
+    # Single-node scenarios
+    problem_size = max_token_num * hidden_dim
+
+    # Large problems (>4M elements): trtllm optimized for throughput
+    if problem_size > 4 * 1024 * 1024:
+        if "trtllm" in suitable_backends:
+            return ["trtllm"]
+
+    # Small token counts (<128): trtllm one-shot has better latency
+    if max_token_num < 128:
+        if "trtllm" in suitable_backends:
+            return ["trtllm"]
+
+    # Small world sizes (<=4): trtllm one-shot efficient
+    if world_size <= 4:
+        if "trtllm" in suitable_backends:
+            return ["trtllm"]
+
+    # Default: return first available
+    return [suitable_backends[0]]
+
+
+# ============================================================================
+# WORKSPACE CREATION - Uses decorator for all validation
+# ============================================================================
+
+
+@backend_requirement(
+    backend_checks={
+        "trtllm": _trtllm_workspace_check,
+        "mnnvl": _mnnvl_workspace_check,
+    },
+    heuristic_func=_workspace_creation_heuristic,
+)
+def create_allreduce_fusion_workspace(
```

Comment:
Could `create_allreduce_fusion_workspace` take an optional `workspace` argument? If `workspace` is big enough or too big this is a noop (maybe just updating backend selection). If it is too small, destroy current workspace and allocate a bigger one.

When we switch to mem pool, we should be able to call `create_allreduce_fusion_workspace` at each forward pass  and memory will just get reused from the mempool (instead of new allocations).
CC @Amir-19